kbs/src/attestation/backend.rs

```diff
@@ -0,0 +1,362 @@
+
+use std::sync::Arc;
+
+use actix_web::{HttpRequest, HttpResponse};
+use anyhow::{anyhow, bail, Context};
+use async_trait::async_trait;
+use base64::{engine::general_purpose::STANDARD, Engine};
+use kbs_types::{Attestation, Challenge, Request, Tee};
+use lazy_static::lazy_static;
+use log::{debug, info};
+use rand::{thread_rng, Rng};
+use semver::{BuildMetadata, Prerelease, Version, VersionReq};
+use serde::Deserialize;
+use serde_json::json;
+
+use crate::attestation::session::KBS_SESSION_ID;
+
+use super::{
+    config::{AttestationConfig, AttestationServiceConfig},
+    session::{SessionMap, SessionStatus},
+    Error, Result,
+};
+
+static KBS_MAJOR_VERSION: u64 = 0;
+static KBS_MINOR_VERSION: u64 = 1;
+static KBS_PATCH_VERSION: u64 = 1;
+
+lazy_static! {
+    static ref VERSION_REQ: VersionReq = {
+        let kbs_version = Version {
+            major: KBS_MAJOR_VERSION,
+            minor: KBS_MINOR_VERSION,
+            patch: KBS_PATCH_VERSION,
+            pre: Prerelease::EMPTY,
+            build: BuildMetadata::EMPTY,
+        };
+
+        VersionReq::parse(&format!("={kbs_version}")).unwrap()
+    };
+}
+
+/// Number of bytes in a nonce.
+const NONCE_SIZE_BYTES: usize = 32;
+
+/// Create a nonce and return as a base-64 encoded string.
+pub async fn make_nonce() -> anyhow::Result<String> {
+    let mut nonce: Vec<u8> = vec![0; NONCE_SIZE_BYTES];
+
+    thread_rng()
+        .try_fill(&mut nonce[..])
+        .map_err(anyhow::Error::from)?;
+
+    Ok(STANDARD.encode(&nonce))
+}
+
+pub(crate) async fn generic_generate_challenge(
+    _tee: Tee,
+    _tee_parameters: serde_json::Value,
+) -> anyhow::Result<Challenge> {
+    let nonce = make_nonce().await?;
+
+    Ok(Challenge {
+        nonce,
+        extra_params: serde_json::Value::String(String::new()),
+    })
+}
+
+/// Interface for Attestation Services.
+///
+/// Attestation Service implementations should implement this interface.
+#[async_trait]
+pub trait Attest: Send + Sync {
+    /// Set Attestation Policy
+    async fn set_policy(&self, _policy_id: &str, _policy: &str) -> anyhow::Result<()> {
+        Err(anyhow!("Set Policy API is unimplemented"))
+    }
+
+    /// Verify Attestation Evidence
+    /// Return Attestation Results Token
+    async fn verify(&self, tee: Tee, nonce: &str, attestation: &str) -> anyhow::Result<String>;
+
+    /// generate the Challenge to pass to attester based on Tee and nonce
+    async fn generate_challenge(
+        &self,
+        tee: Tee,
+        tee_parameters: serde_json::Value,
+    ) -> anyhow::Result<Challenge> {
+        generic_generate_challenge(tee, tee_parameters).await
+    }
+}
+
+/// Attestation Service
+#[derive(Clone)]
+pub struct AttestationService {
+    /// Attestation Module
+    inner: Arc<dyn Attest>,
+
+    /// A concurrent safe map to keep status of RCAR status
+    session_map: Arc<SessionMap>,
+
+    /// Max timeout between `auth` and `attest` request of a client
+    timeout: i64,
```

Comment:
in #162 I raised the question about this `timeout` and IIRC @fitzthum mentioned it somewhere too. I'm asking the question again since we are making big config changes: do we need to have this configurable for KBS or should it just follow the token `exp`.

With the current CDH setup, is it possible, for example to set 1 minute here but CDH can continue to pull secrets as long as the token is valid (5min by default)?

Reply:
This timeout as noted here is not the expiration of the token. It is just the maximum delay between an `auth` request and an `attest` request. That means if the `attest` comes too late after `auth`, the RCAR handshake will fail. You raised a good idea about the inner `timeout` for session. I think we should distinguish the two different `timeout`s. Btw, do you think we need to add another option to specify the timeout of the session time, or use a hardcode one? cc @fitzthum 

> With the current CDH setup, is it possible, for example to set 1 minute here but CDH can continue to pull secrets as long as the token is valid (5min by default)?

Actaully it can for now : |  The CDH uses the token (via HTTP Auth header of request) to pull secrets rather than reuse the session.

Reply:
This is a little tricky. Actually the both timeout are the session timeout. What it will influence is when a secret access comes, the token will firstly be looked up from the session. If the session is expired, the http auth header will be checked to get the token.

As for whether the resource retrievement is allowed, it is the checking against the token that matters. Some different cases
1. timeout > token expiration time: the token can be got. When checking against the token, kbs will realize the token is expired then failed to pull secret.
2. timeout < token expiration time: when trying to get token from session, the kbs will realize the session is expired. Then try to get token from http auth header. If a token is found, the secret pulling will succeed. If no tokens are found, the secret pulling will fail.

Reply:
Yup. I added a commit for Token verifier to check the expiration time of the token. Hopefully things can be clear and we can discuss more back to #162

Reply:
that is not necessary since jsonwebtoken handles that [automatically when using the default `Validation`](https://github.com/confidential-containers/trustee/blob/96ddb44308a7351baa83f6f755350babe285578c/kbs/src/token/jwk.rs#L213). It's also possible to make that explicitly.

Reply:
Oh yes. nice. Let me delete the last commit